packages/autocomplete-core/src/types/plugins.ts

```diff
@@ -0,0 +1,17 @@
+import { AutocompleteOptions, AutocompleteSource } from './api';
+
+export type AutocompletePlugin<TItem, TData = unknown> = Partial<
+  Pick<AutocompleteOptions<TItem>, 'onStateChange' | 'onSubmit' | 'getSources'>
+> & {
+  /**
+   * The subscribed properties are properties that are called when other sources
+   * are interacted with.
+   */
+  subscribed: {
+    onSelect: AutocompleteSource<TItem>['onSelect'];
+  };
+  /**
+   * An extra plugin specific object to store variables and functions
+   */
+  data: TData;
```

Comment:
good call here, in InstantSearch we see that keeping state for a single widget is complicated